presto-main/src/main/java/com/facebook/presto/operator/ExchangeClient.java

```diff
@@ -165,11 +170,36 @@ public synchronized void addLocation(URI location)
                 scheduler,
                 pageBufferClientCallbackExecutor);
         allClients.put(location, client);
+        taskIdToURI.put(sourceTaskId, location);
         queuedClients.add(client);
 
         scheduleRequestIfNecessary();
     }
 
+    public synchronized void removeRemoteSource(TaskId sourceTaskId)
+    {
+        requireNonNull(sourceTaskId, "sourceTaskId is null");
+
+        // Ignore remove remote source after close
+        if (closed.get()) {
+            return;
+        }
+
+        URI location = taskIdToURI.get(sourceTaskId);
+        if (location == null) {
+            return;
+        }
+
+        HttpPageBufferClient client = allClients.get(location);
+        if (client == null) {
+            return;
+        }
+
+        client.close();
```

Comment:
Can `close` throw here? If yes, would that be a problem if we don't add the `client` to below collections in that case?

Reply:
@nezihyigitbasi  Good catch! I switch to use `closeQuietly(client)`